Using Clang __builtin_assume if available.

verri opened the pull request:
Clang must be verified first because it also defines `__GNUC__`.

`__builtin_assume` is a recent addition (since 2014), so I also check if it is available.

verri committed:
Using Clang __builtin_assume if available.

foonathan commented on debug_assert.hpp:
```diff
@@ -44,7 +44,18 @@
-#ifdef __GNUC__
+#ifdef __clang__
+#if __has_builtin(__builtin_assume)
+#define DEBUG_ASSERT_ASSUME(Expr) __builtin_assume(Expr)
```
Can you use `defined(__clang__) && __has_builtin...` there and then just have a single `#elif defined(__GNUC__)`? Avoids the repetition.

verri commented:
That was my first tentative, but gcc does not compile.

foonathan commented:
Ah, I see. Could you add a comment then?

verri commented:
Where? Explaining the need for the repetition?

foonathan commented:
Yes.

verri commented:
Another possibility is:

```c++
// Checking for clang must come first because clang also defines __GNUC__.
#ifndef DEBUG_ASSERT_ASSUME
#ifdef __clang__
// __has_builtin may not work in other compilers.
#if __has_builtin(__builtin_assume)
#define DEBUG_ASSERT_ASSUME(Expr) __builtin_assume(Expr)
#endif
#endif
#endif

#ifndef DEBUG_ASSERT_ASSUME
#ifdef __GNUC__
#define DEBUG_ASSERT_ASSUME(Expr)                                                                  \
    do                                                                                             \
    {                                                                                              \
        if (!(Expr))                                                                               \
            __builtin_unreachable();                                                               \
    } while (0)
#elif defined(_MSC_VER)
#define DEBUG_ASSERT_ASSUME(Expr) __assume(Expr)
#else
/// Hint to the compiler that a condition is `true`.
/// Define it yourself prior to including the header to override it.
#define DEBUG_ASSERT_ASSUME(Expr)
#endif
#endif
```

Do you think it is better?

verri commented:
I realized something that might be a problem. Compiling example.cpp with clang now emits this warning:

```
In file included from example.cpp:1:
./debug_assert.hpp:240:33: warning: the argument to '__builtin_assume' has side effects that will be discarded [-Wassume]
            DEBUG_ASSERT_ASSUME(expr());
                                ^~~~~~
./debug_assert.hpp:51:52: note: expanded from macro 'DEBUG_ASSERT_ASSUME'
#define DEBUG_ASSERT_ASSUME(Expr) __builtin_assume(Expr)
                                                   ^~~~
example.cpp:18:5: note: in instantiation of function template specialization 'debug_assert::detail::do_assert<(lambda at
      example.cpp:18:5), module_a, 2>' requested here
    DEBUG_ASSERT(2 + 2 == 4, module_a{}, debug_assert::level<2>{}); // assertion with level
    ^
./debug_assert.hpp:293:27: note: expanded from macro 'DEBUG_ASSERT'
    debug_assert::detail::do_assert([&] { return Expr; }, DEBUG_ASSERT_CUR_SOURCE_LOCATION, #Expr, \
                          ^
1 warning generated.
```

I think clang is not able to understand that Expr does not have side effects.

foonathan commented:
I don't think it's a good idea then.

verri commented:
Does MSC emit warnings?

foonathan commented:
I'm afraid I haven't checked thoroughly  but I haven't noticed any.

verri committed:
Fixed warning in clang.

- Added an attribute to indicate that the assertion expression
is pure (as it should be).

verri commented:
Marking the lambda as a pure function (as it should be) solves the warnings.
